source/extensions/filters/http/set_metadata/set_metadata_filter.cc

```diff
@@ -0,0 +1,83 @@
+#include "extensions/filters/http/set_metadata/set_metadata_filter.h"
+
+#include "envoy/extensions/filters/http/set_metadata/v3/set_metadata.pb.h"
+
+#include "common/config/well_known_names.h"
+#include "common/http/utility.h"
+#include "common/protobuf/protobuf.h"
+
+#include "extensions/filters/http/well_known_names.h"
+
+#include "absl/strings/str_format.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace SetMetadataFilter {
+
+namespace {
+
+void pbStructUpdate(ProtobufWkt::Struct& obj, ProtobufWkt::Struct const& with) {
+  auto& objfields = *obj.mutable_fields();
+
+  for (auto const& [key, val] : with.fields()) {
+    auto& objkey = objfields[key];
+    switch (val.kind_case()) {
+    // For scalars, the last one wins.
+    case ProtobufWkt::Value::kNullValue:
+    case ProtobufWkt::Value::kNumberValue:
+    case ProtobufWkt::Value::kStringValue:
+    case ProtobufWkt::Value::kBoolValue:
+      objkey = val;
+      break;
+    // If we got a structure, recursively update.
+    case ProtobufWkt::Value::kStructValue:
+      pbStructUpdate(*objkey.mutable_struct_value(), val.struct_value());
+      break;
+    // For lists, append the new values.
+    case ProtobufWkt::Value::kListValue: {
+      auto& objkeyvec = *objkey.mutable_list_value()->mutable_values();
+      auto& vals = val.list_value().values();
+      objkeyvec.MergeFrom(vals);
```

Comment:
I'm not sure exactly how `MergeFrom` and standard `Struct` merging works. In https://github.com/envoyproxy/envoy/issues/16266#issuecomment-832929796 it's claimed it does the wrong thing and so we need a custom merger here. Have you validated this empirically?

I think this merger should be a library if needed and placed in `source/common/protobuf`, since we probably want this as a general `StructUtils` kind of thing. But I can't tell from reading the description of `MergeFrom` if it doesn't already do what you need.

Reply:
I tried `MergeFrom`, and indeed validated empirically that it does not do what we want *for structures*, as per the example you mentionned. For lists, it is just appending the values (which is what we want), as we can see here https://github.com/protocolbuffers/protobuf/blob/465acaab6651a158592b06605a38fed373512f23/src/google/protobuf/struct.pb.cc#L1023 and here https://github.com/protocolbuffers/protobuf/blob/465acaab6651a158592b06605a38fed373512f23/src/google/protobuf/repeated_field.h#L1826-L1867 . I also validated this empirically and indirectly in the unit tests. I also couldn't find any proper documentation on this...

I'm okay to move this into `source/common/protobuf`, with a separate commit and unit tests!

Reply:
OK, let's do that then. This kind of protobuf merge is probably of interest to @envoyproxy/api-shepherds; do we want to be using this as a standard merge feature in the APIs when combining Structs?

Reply:
So commit https://github.com/envoyproxy/envoy/pull/16400/commits/acb5e25389b4879cbc328dd85464689f6be9b616 introduce `StructUtil::update`, with the related tests. And commit https://github.com/envoyproxy/envoy/pull/16400/commits/7540c2a07e1a8d4f9ac70006d6ffda037c051e87 uses it for the set-metadata filter.

Writing the tests for `StructUtil::update` made me realize that I didn't take into account the case where two keys exist in the original and new object but do not have the same type. I took the strategy of "last-one-win" here (that is taking the value of the new object). I modified the documentation of set-metadata to also discuss this case.

I also just realized I forgot the doc of the `StructUtil::update` API itself, I'm writing a commit to add it.

BTW, when I'll rebase/squash I'll obviously put the commit about `StructUtil::update` first in history!

Reply:
API documentation is in https://github.com/envoyproxy/envoy/pull/16400/commits/4e846f0d5fe4dc9551ea951947d0337841c2b9d6